src/celeritas/field/RZPhiMapFieldData.hh

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/field/RZPhiMapFieldData.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/cont/Span.hh"
+#include "corecel/data/Collection.hh"
+#include "corecel/math/Turn.hh"
+
+#include "FieldDriverOptions.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * FieldMap (3-dimensional RZ-Phi map) grid data
+ */
+template<Ownership W, MemSpace M>
+struct RZPhiMapGridData
+{
+    template<class T>
+    using Items = Collection<T, W, M>;
+    Items<real_type> z;
+    Items<real_type> r;
+    Items<Turn> phi;
+
+    operator bool() const { return !z.empty() && !r.empty() && !phi.empty(); }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * FieldMap element for RZ-Phi map
+ */
+struct RZPhiMapElement
+{
+    real_type value_z;
+    real_type value_r;
+    real_type value_phi;
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Device data for interpolating field values.
+ */
+template<Ownership W, MemSpace M>
+struct RZPhiMapFieldParamsData
+{
+    //! Grids of FieldMap
+    RZPhiMapGridData<W, M> grids;
+
+    //! Options for FieldDriver
+    FieldDriverOptions options;
+
+    //! Index of FieldMap Collection
+    using ElementId = ItemId<size_type>;
+
+    template<class T>
+    using ElementItems = Collection<T, W, M, ElementId>;
+    ElementItems<RZPhiMapElement> fieldmap;
+
+    //! Check whether the data is assigned
+    explicit inline CELER_FUNCTION operator bool() const
+    {
+        return !fieldmap.empty();
+    }
+
+    inline CELER_FUNCTION bool
+    valid(real_type z, real_type r, real_type phi) const
+    {
+        CELER_EXPECT(grids);
+        Turn turn_phi{phi / Turn::unit_type::value()};
+        auto view_z
+            = Span<real_type const>{grids.z.data().get(), grids.z.size()};
+        auto view_r
+            = Span<real_type const>{grids.r.data().get(), grids.r.size()};
+        auto view_phi
+            = Span<Turn const>{grids.phi.data().get(), grids.phi.size()};
+        return (z >= view_z.front() && z <= view_z.back()
+                && r >= view_r.front() && r <= view_r.back()
+                && turn_phi >= view_phi.front() && turn_phi <= view_phi.back());
+    }
+
+    inline CELER_FUNCTION ElementId id(size_type idx_z,
+                                       size_type idx_r,
+                                       size_type idx_phi) const
+    {
+        CELER_EXPECT(grids);
+        // Index with ordering [Z][R][Phi]
+        return ElementId(idx_z * grids.r.size() * grids.phi.size()
+                         + idx_r * grids.phi.size() + idx_phi);
+    }
```

Comment:
Should we replace this with the `HyperslabIndexer` (and same in `RZMap`)?

Reply:
There are a bunch of places where we do this and yes please do. Also, I know RZ map indexes as [Z][R] which I didn't realize during the original pull, but let's keep the accessors C-like: [Phi][R][Z].